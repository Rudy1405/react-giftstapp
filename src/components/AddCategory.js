import React, {useState} from 'react'
import PropTypes from "prop-types"

export const AddCategory = ( {setCategories} ) => { // aqui como se hizo q en el padre mandara algo aca lo cachamos y asi lo mandamos a llmar

    const [inputValue, setInputValue] = useState("Add a category");
    const handleInputChange = (element) => {
        setInputValue(element.target.value); //Tras hacer esto el INPUTVALUE tendra el ultimo valor escrito
    }
    const handleSubmit = (element) => {
        element.preventDefault(); /// ****** Esto es para evitar que el form provoque un refresh de la pagina
        if(inputValue.trim().length > 2){
            setCategories( categories =>  [ inputValue, ...categories] ); // obt el estado anterior y anexamos el input value
            setInputValue("");
        }
    }
    const handleClick = (element) => {
        setInputValue("");
    }
    /// en este caso no usamos el <></> para indicar que hay un elemento agrupador del componente porque el FORM ya lo es
    return (
          <form className="animate__animated animate__zoomIn animate__delay-0.5s" onSubmit={handleSubmit}>
              <p> {inputValue} </p>
            <input 
                    type="text"
                    value= {inputValue}
                    onChange={ handleInputChange }
                    onClick={handleClick}

            />
          </form>
    )
}

AddCategory.propTypes = {
        setCategories: PropTypes.func.isRequired  // es func porque el setCategories es una funcion lo que se envia
}

AddCategory.defaultProps = {
    setCategories: []
}
