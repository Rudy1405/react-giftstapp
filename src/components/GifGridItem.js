import { prototype } from 'enzyme-adapter-react-16'
import React from 'react'
import PropTypes from 'prop-types'; // shortcut con impt

export const GifGridItem = ( {id,title,url} ) => { // aqui ya tenemos los props enviados por [...img] por ende podemos llamar asi a sus elementos solo para usarlos directamente cada vez que se cree un componente de estos
   
    return (
        <div className="card animate__animated animate__fadeInUp animate__delay-0.5s">
             <img src={url? url : ""} alt={title? title : "No Title"} />
             <p>{title? title : "Imagen Generica"}</p>
        </div>
    )
}

GifGridItem.protoType = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
}