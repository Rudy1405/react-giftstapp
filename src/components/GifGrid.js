import React, {useState, useEffect} from 'react'
import PropTypes from 'prop-types';
//import { getGifs } from '../helpers/getGifs';
import { useFetchGif } from '../hooks/useFetchGif';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => {


    const {data:images, loading} = useFetchGif(category); // como usefetchGif regresa un obj, lo podemos desestructurar
          /// en la desesctructracion podemos renombrar variables, por ejemplo data llamarle images para usarlo como images abajo
/*
    El useEffect nos servia para cargar por primera vez todo, pero se uso el customHOOK de usefetchgif por ende ya no se ocupara

    const [imagesValue, setImagesValue] = useState([])
    useEffect( ()=>{ // useEffect es un hook para ejecutar codigo de forma condicional para que el getgifs por cualqueir cambio no se este llame y llame pueh
        getGifs(category) // mando a llamar a mi helper que es una promesa que me hace el fetch de lo que mande en category desde el AddCategory
        .then(imgs => setImagesValue(imgs)) 
    }, [category]) // aqui van las dependencias de cuando ejecutar el codigo... un [] es para que solo se ejecute 1 vez
                   // se pone con [category] para evitar un warning de que el category puede cambiar en alguna de sus posiciones 
     /// antes aqui estaba el getgifs pero se paso a un helper para tener mejor acomodado el codigo, ahora se llama drecto en el useEffect
*/
   



    return (
        <>
            <h3 className="animate__animated animate__fadeInUp"> {category} </h3>

            <div className="container">
                <div className="row">
                    {loading && <p className="animate__animated animate__hinge animate__delay-0.5s">Loading...</p>}    { /*  lo anterior es igual a {loading ? <p>Loading...</p> : null} */}
                    <div className='class="col-6"' className="cardGrid">            
                        { // las {} es para denotar que vamos a meter JS
                            images.map( img => { // en lugar de (img) se puede desestructurar a {id,title,url} para solo tener los elementos POR elemento que queremos
                            return <GifGridItem 
                                            key= {img.id}
                                            {...img } // mandandolo asi, estamos mandando solo las propiedades de cada IMG y por ende poder desestructurar en el GIFGRIDITEM
                                /> // Creamos un nuevo componente porque pues adentro ira un pedazo de HTML con mucho mas cuerpo
                            })  
                        }
                    </div>
                </div>
            </div>

            { // Esto es con la version del useEfecct antes del customHook
            /* <div className="container"> 
                <div className="row">
                    <div className='class="col-6"' className="cardGrid">            
                        { // las {} es para denotar que vamos a meter JS
                            imagesValue.map( img => { // en lugar de (img) se puede desestructurar a {id,title,url} para solo tener los elementos POR elemento que queremos
                            return <GifGridItem 
                                            key= {img.id}
                                            {...img } // mandandolo asi, estamos mandando solo las propiedades de cada IMG y por ende poder desestructurar en el GIFGRIDITEM
                                /> // Creamos un nuevo componente porque pues adentro ira un pedazo de HTML con mucho mas cuerpo
                            })  
                        }
                    
                    </div>
                </div>
            </div> */}
           
            
        </>
    )
}


GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}