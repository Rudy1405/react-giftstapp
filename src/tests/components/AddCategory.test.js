import React from "react"
import "@testing-library/jest-dom";
import { AddCategory } from "../../components/AddCategory"
import { shallow } from "enzyme"

describe('Pruebas en <AddCategory />', () => {

    const setCategories = jest.fn(); //sirve como una referencia a una funcion
    let wrapper = shallow(<AddCategory setCategories={setCategories}/>);

    beforeEach(()=>{ //ciclo de vida de pruebas entonces lo de adentro se ejecutara CADA QUE se haga un test
        jest.clearAllMocks(); // limpia cualquier data mockeada que se haya declarado
        wrapper = shallow(<AddCategory setCategories={setCategories}/>); //lo inicializamos dos veces solo para que al codear aparezcan las ayudas como wrapper.f dara .find :) 
    });

    test('should show correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('should change input text', () => {
        const input = wrapper.find("input");
        const value = "Hola Mundo"; //simulamos el e.target.value del EVENTO para que el componente lo entienda.
        input.simulate("change", { target: {value}}); //es change mas no ONCHANGE porque es un input
        expect( wrapper.find("p").text().trim() ).toBe(value); // Se creo un P en HTML para aqui ver que el value que se mete es lo que se displaa y se usa en la logica

    })

    test('should not POST info  OnSubmit', () => {
        wrapper.find("form").simulate("submit", { preventDefault: ()=>{}}); //simular el ONSUBMIT DEL FORM, el preventDefault es una funcion que tiene el handleonsubmit (en el archivo original) por ende necesita reconocerlo  puede ser asi o asi preventDefault(){} que es una func vacia igual
        // como no hay un setCategory no deberia hacer nada 
        expect(setCategories).toHaveBeenCalled(); // setCategory si fue llamado pero si valor es el por deafult linia 6 se usa el useState para dar un default Add a category 
        expect(wrapper.find("input").props().value).toBe(""); //aqui debe ser vacio porque el useState pone un valor visual nomas, eso NO se envia si se hace enter o submit en el form
    })

    test('should call SetCategory and wipe text input', () => {
        const input = wrapper.find("input");
        const value = "chanclas";
        input.simulate("change", { target: {value}}); //simulamos que se tiene algo en el text input
        wrapper.find("form").simulate("submit", { preventDefault: ()=>{}}); // simulamos el enter para mandar el value 
        // para este entonces ya se mando la data y el input text deberia estar vacio, bueno no, por el USESTATE linea 6 deberia ser el texto Add a category que es el default
        expect(setCategories).toHaveBeenCalledTimes(1); // que se haya llamado solo 1 vez porque fue 1 input
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function)); // y que aparte el argumento sea una funcion
        expect(input.props().value).toBe("Add a category"); //aqui no hacemos TRIM del value porque este es el por default y de sebe respetar ALV
    })
    

    
    
    
})
