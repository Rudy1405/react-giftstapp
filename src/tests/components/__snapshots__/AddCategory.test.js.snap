// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> should show correctly 1`] = `
<form
  className="animate__animated animate__zoomIn animate__delay-0.5s"
  onSubmit={[Function]}
>
  <p>
     
    Add a category
     
  </p>
  <input
    onChange={[Function]}
    onClick={[Function]}
    type="text"
    value="Add a category"
  />
</form>
`;
